### DISCUSSION TITLE
Adding attributes existing typed class (eg. custom pytest markers)

### DISCUSSION THREAD

**Discussion by ljodal** [microsoft/pyright#5732] [Q&A]
Hi, I'm wondering if there's any way to let pyright know about additional attributes on an existing (typed) object/class. Specifically my use case right now is that I'm defining some custom pytest markers that I'd like to let pyright know about to provide autocompletion and type hints.

For the built-in markers pytest "cheats" by defining markers attributes [in a `TYPE_CHECKING` block](https://github.com/pytest-dev/pytest/blob/main/src/_pytest/mark/structures.py#L513-L520). Basically I'd like to do the same for the custom markers we have, but without having to redefine the types for all of the library/module.

I'm not too concerned about actual type checking here, but having autocomplete work would be very useful to help with discovery of the custom markers in our project.

---

For those unfamiliar with pytest markers, it's a decorator you can add to your test functions which basically records the arguments and makes them available to e.g. a fixture. For example something like this:

```python
import pytest
from django.utils import translation

@pytest.fixture(autouse=True)
def activate_translation(request: pytest.FixtureRequest) -> Iterable[None]:
    """
    This fixture works like a context manager,
    which will be automatically wrapped around the test
    """
    language = "en"
    if marker := request.node.get_closest_marker("language"):
         language = marker.args[0]
    
    with translation.override(language):
        yield  # <-- Test function is called when we yield here

def test_default_value() -> None:
    assert translation.get_language() == "en"

@pytest.mark.language("nb")
def test_override_with_mark() -> None:
    assert translation.get_language() == "nb"
```

In this example I'd like to tell pyright that the signature of `pytest.mark.langauge` is something like this:

```python
F = TypeVar("F", bound=Callable[..., Any])
def language(language: str, /) -> Callable[[F], F]: ...
```

**Comment by erictraut** [comment]
The only solution I can think of is to create type stubs for `pytest` (e.g. using mypy's stubgen or pyright's `--createstub` command-line option) and then manually add in the type information for your custom marks. These stubs would then be placed in a local `typings` directory within your project. These stubs would override the type information in the actual `pytest` library according to the ordering rules established in PEP 561. The downside is that you would need to maintain these stubs and update them if pytest changes.

**Comment by ljodal** [comment]
Hmm, okay, thanks for the answer! I'd rather not maintain a full set of stubs, so that's kinda out of the question. For mypy I can fairly easily write a plugin for this, but seeing as most people (at our company at least) uses pyright through their IDEs that wouldn't really help much.

What would it take to get support for some way of doing this? In short I'm thinking of some way to extend the types of an existing project, which I guess could be useful for other kinds of plugins that extend something as well. Not entirely sure what that would look like, but I envision some way of writing a `.pyi` file to patch, rather than overwire an existing module. I don't have any other use cases in front of me, but this would at least be useful for all pytest plugins that provide custom marks. I imagine there's other similar cases as well.

Maybe a module along these lines that can be exposed as a setuptools entrypoint and loaded by the type checker?

```python
# stubs.pyi

from typing import Any, Callable, Protocol, TypeVar
import pytest

F = TypeVar("F", bound=Callable[..., Any])

class MyPytestMarker(Protocol):
    def __call__(self, language: str, /) -> Callable[[F], F]: ...

# Define the type of the attribute
pytest.mark.language: MyPytestMarker
```

The last line there would be the magic here. That statement doesn't seem to much now. Pyright seems to ignore it and mypy produces two errors:

```
error: Type cannot be declared in assignment to non-self attribute
error: "MarkGenerator" has no attribute "language"
```

**Reply by erictraut** [reply]
If you primarily want this functionality for completion suggestions and signature help and don't need type checking (i.e. errors generated if you use incorrect arguments), then you could try to convince the pylance team to add support for custom pytest markers in the pylance completion and signature help providers. They've already added some support for pytest fixtures, so they might be open to extending this.

If you want type checking support, I think you have the following options:
1. Create a fork of pyright and build (and maintain) your own custom version that supports the functionality you're looking for.
2. Propose and champion an extension to the Python type system that addresses your use case. This would probably require authoring a new PEP and getting it approved by the typing community and the Python steering council. Pyright is a standards-based type checker, and it supports all the typing standards (and often serves as the reference implementation for PEPs that are still in the draft stage). If you want to pursue this, I recommend starting a new discussion thread in the python/typing project.
3. Switch to mypy and write a custom mypy plugin.

**Reply by ljodal** [reply]
Okay, thanks for the answer! My primary concern is auto-complete, and mostly not for myself but the rest of the developers in the company that I'm making tooling for. I think most use VSCode with pyright/pylance, but when I think about it there's probably a lot of pycharm users as well. So bot 1 and 3 are not really options, I guess 2 might work, but that seems like a long process so I guess we'll just have to manage without for now.

I'll let it sit in the back of my mind and see if I come up with better alternatives. I'll bring it to python/typing if I feel like I have a good solution.

Once again, thanks for you quick and thorough responses! 🙇
